File: traffic_light.js
/**
 * 사거리 신호등 시스템 - JavaScript 상태 머신
 *
 * 단계(Phase) 순환:
 *   NS_GREEN  → NS_YELLOW  → EW_GREEN  → EW_YELLOW  → (반복)
 *   남북 초록     남북 황색     동서 초록     동서 황색
 */

// ─── 신호 단계 정의 ───────────────────────────────────────────
const PHASES = {
  NS_GREEN:  'NS_GREEN',
  NS_YELLOW: 'NS_YELLOW',
  EW_GREEN:  'EW_GREEN',
  EW_YELLOW: 'EW_YELLOW',
};

// 각 단계의 신호 상태 매핑
//   각 방향: 'red' | 'yellow' | 'green'
const PHASE_MAP = {
  [PHASES.NS_GREEN]: {
    north: 'green',  south: 'green',
    east:  'red',    west:  'red',
    pedNS: 'stop',   pedEW: 'walk',   // 보행자: NS도로 정지, EW보도 통행
    label: '남북 직진 (초록)',
  },
  [PHASES.NS_YELLOW]: {
    north: 'yellow', south: 'yellow',
    east:  'red',    west:  'red',
    pedNS: 'stop',   pedEW: 'stop',
    label: '남북 주의 (황색)',
  },
  [PHASES.EW_GREEN]: {
    north: 'red',    south: 'red',
    east:  'green',  west:  'green',
    pedNS: 'walk',   pedEW: 'stop',   // 보행자: NS보도 통행, EW도로 정지
    label: '동서 직진 (초록)',
  },
  [PHASES.EW_YELLOW]: {
    north: 'red',    south: 'red',
    east:  'yellow', west:  'yellow',
    pedNS: 'stop',   pedEW: 'stop',
    label: '동서 주의 (황색)',
  },
};

// 단계 순서
const PHASE_ORDER = [
  PHASES.NS_GREEN,
  PHASES.NS_YELLOW,
  PHASES.EW_GREEN,
  PHASES.EW_YELLOW,
];

// ─── 상태 ──────────────────────────────────────────────────────
let state = {
  running:       false,
  phaseIndex:    0,          // PHASE_ORDER 인덱스
  timeLeft:      0,          // 남은 초
  cycleCount:    0,
  speed:         1,          // 배속 (1~10)
  intervalId:    null,
  durations: {
    nsGreen:  30,
    nsYellow:  5,
    ewGreen:  30,
    ewYellow:  5,
  },
};

// ─── 단계별 지속 시간 조회 ──────────────────────────────────────
function getPhaseDuration(phaseKey) {
  switch (phaseKey) {
    case PHASES.NS_GREEN:  return state.durations.nsGreen;
    case PHASES.NS_YELLOW: return state.durations.nsYellow;
    case PHASES.EW_GREEN:  return state.durations.ewGreen;
    case PHASES.EW_YELLOW: return state.durations.ewYellow;
  }
}

// ─── DOM 헬퍼 ──────────────────────────────────────────────────
function el(id) { return document.getElementById(id); }

function setLightState(direction, signalColor) {
  ['red', 'yellow', 'green'].forEach(color => {
    const node = el(`${direction}-${color}`);
    if (node) {
      node.classList.toggle('on', color === signalColor);
    }
  });
}

function setPedestrianState(side, pedState) {
  // side: 'ns' or 'ew'
  // pedState: 'walk' | 'stop' | 'wait'
  const dirs = side === 'ns' ? ['north', 'south'] : ['east', 'west'];
  dirs.forEach(dir => {
    const icon  = el(`ped-${dir}-icon`);
    const label = el(`ped-${dir}-label`);
    if (!icon || !label) return;

    icon.className  = `ped-icon ${pedState}`;
    label.className = `ped-label ${pedState}`;

    if (pedState === 'walk') {
      icon.textContent  = '🚶';
      label.textContent = '통행';
    } else if (pedState === 'stop') {
      icon.textContent  = '🚷';
      label.textContent = '정지';
    } else {
      icon.textContent  = '🚶';
      label.textContent = '대기';
    }
  });
}

function highlightPhaseStep(phaseKey) {
  const stepMap = {
    [PHASES.NS_GREEN]:  'step-ns-green',
    [PHASES.NS_YELLOW]: 'step-ns-yellow',
    [PHASES.EW_GREEN]:  'step-ew-green',
    [PHASES.EW_YELLOW]: 'step-ew-yellow',
  };
  Object.values(stepMap).forEach(id => {
    const node = el(id);
    if (node) node.classList.remove('active');
  });
  const active = el(stepMap[phaseKey]);
  if (active) active.classList.add('active');
}

function updateCarAnimation(phaseKey) {
  const carNS = el('car-ns');
  const carEW = el('car-ew');
  if (!carNS || !carEW) return;

  const nsMoving = phaseKey === PHASES.NS_GREEN;
  const ewMoving = phaseKey === PHASES.EW_GREEN;

  carNS.classList.toggle('running', nsMoving);
  carEW.classList.toggle('running', ewMoving);
}

// ─── 단계 적용 ──────────────────────────────────────────────────
function applyPhase(phaseKey) {
  const cfg = PHASE_MAP[phaseKey];
  if (!cfg) return;

  // 신호등 업데이트
  ['north', 'south', 'east', 'west'].forEach(dir => {
    setLightState(dir, cfg[dir]);
  });

  // 보행자 신호 업데이트
  setPedestrianState('ns', cfg.pedNS);
  setPedestrianState('ew', cfg.pedEW);

  // 타임라인 하이라이트
  highlightPhaseStep(phaseKey);

  // 차량 애니메이션
  updateCarAnimation(phaseKey);

  // 상태 표시
  el('phase-name').textContent = cfg.label;
}

// ─── 초기 상태 렌더링 ──────────────────────────────────────────
function renderInitialState() {
  // 모든 신호등 끄기
  ['north', 'south', 'east', 'west'].forEach(dir => {
    ['red', 'yellow', 'green'].forEach(color => {
      const node = el(`${dir}-${color}`);
      if (node) node.classList.remove('on');
    });
  });

  // 보행자 대기
  ['north', 'south', 'east', 'west'].forEach(dir => {
    const icon  = el(`ped-${dir}-icon`);
    const label = el(`ped-${dir}-label`);
    if (icon)  { icon.textContent  = '🚶'; icon.className  = 'ped-icon'; }
    if (label) { label.textContent = '대기'; label.className = 'ped-label'; }
  });

  el('phase-name').textContent   = '-';
  el('countdown').textContent    = '--';
  el('cycle-count').textContent  = '0';

  // 타임라인 초기화
  ['step-ns-green','step-ns-yellow','step-ew-green','step-ew-yellow'].forEach(id => {
    const node = el(id);
    if (node) node.classList.remove('active');
  });

  // 차량 정지
  const carNS = el('car-ns');
  const carEW = el('car-ew');
  if (carNS) carNS.classList.remove('running');
  if (carEW) carEW.classList.remove('running');

  updateDurationLabels();
}

// ─── 틱 (매 초 호출) ──────────────────────────────────────────
function tick() {
  if (!state.running) return;

  state.timeLeft -= 1;
  el('countdown').textContent = Math.max(0, state.timeLeft);

  if (state.timeLeft <= 0) {
    // 다음 단계로
    state.phaseIndex = (state.phaseIndex + 1) % PHASE_ORDER.length;

    // 사이클 카운트: NS_GREEN 으로 돌아올 때마다 +1
    if (state.phaseIndex === 0) {
      state.cycleCount += 1;
      el('cycle-count').textContent = state.cycleCount;
    }

    const nextPhase = PHASE_ORDER[state.phaseIndex];
    state.timeLeft  = getPhaseDuration(nextPhase);
    applyPhase(nextPhase);
    el('countdown').textContent = state.timeLeft;
  }
}

// ─── 시뮬레이션 제어 ──────────────────────────────────────────
function startSimulation() {
  if (state.running) return;
  state.running = true;
  el('btn-start-stop').textContent = '일시정지';

  // 처음 시작 시 초기 단계 설정
  if (state.timeLeft <= 0) {
    state.phaseIndex = 0;
    const firstPhase = PHASE_ORDER[0];
    state.timeLeft   = getPhaseDuration(firstPhase);
    applyPhase(firstPhase);
    el('countdown').textContent = state.timeLeft;
  }

  const intervalMs = Math.round(1000 / state.speed);
  state.intervalId = setInterval(tick, intervalMs);
}

function pauseSimulation() {
  if (!state.running) return;
  state.running = false;
  el('btn-start-stop').textContent = '계속';
  clearInterval(state.intervalId);
  state.intervalId = null;
}

function toggleSimulation() {
  if (state.running) {
    pauseSimulation();
  } else {
    startSimulation();
  }
}

function resetSimulation() {
  const wasRunning = state.running;
  pauseSimulation();

  state.phaseIndex  = 0;
  state.timeLeft    = 0;
  state.cycleCount  = 0;

  el('btn-start-stop').textContent = '시작';
  renderInitialState();

  if (wasRunning) {
    // 리셋 후 자동 재시작
    startSimulation();
  }
}

// ─── 속도 변경 ──────────────────────────────────────────────────
function changeSpeed(value) {
  state.speed = parseInt(value, 10);
  el('speed-label').textContent = `${state.speed}x`;

  if (state.running) {
    // 인터벌 재시작
    clearInterval(state.intervalId);
    const intervalMs = Math.round(1000 / state.speed);
    state.intervalId = setInterval(tick, intervalMs);
  }
}

// ─── 지속 시간 업데이트 ────────────────────────────────────────
function updateDuration(key, value) {
  const parsed = parseInt(value, 10);
  if (isNaN(parsed) || parsed < 1) return;
  state.durations[key] = parsed;
  updateDurationLabels();
}

function updateDurationLabels() {
  el('dur-ns-green').textContent  = `${state.durations.nsGreen}s`;
  el('dur-ns-yellow').textContent = `${state.durations.nsYellow}s`;
  el('dur-ew-green').textContent  = `${state.durations.ewGreen}s`;
  el('dur-ew-yellow').textContent = `${state.durations.ewYellow}s`;
}

// ─── 키보드 단축키 ────────────────────────────────────────────
document.addEventListener('keydown', (e) => {
  if (e.code === 'Space') {
    e.preventDefault();
    toggleSimulation();
  } else if (e.code === 'KeyR') {
    resetSimulation();
  }
});

// ─── 초기화 ──────────────────────────────────────────────────
renderInitialState();

File: traffic_light.py
#!/usr/bin/env python3
"""
사거리 신호등 시스템 - Python CLI 시뮬레이터

사용법:
    python traffic_light.py              # 기본 실행
    python traffic_light.py --speed 2   # 2배속 실행
    python traffic_light.py --cycles 3  # 3사이클 후 종료
    python traffic_light.py --ns 45 --ew 30  # 시간 직접 지정
"""

import time
import sys
import argparse
from enum import Enum
from dataclasses import dataclass, field
from typing import Dict


# ─── 신호 색상 ────────────────────────────────────────────────
class Signal(Enum):
    RED    = "red"
    YELLOW = "yellow"
    GREEN  = "green"


# ─── 단계 ────────────────────────────────────────────────────
class Phase(Enum):
    NS_GREEN  = "NS_GREEN"
    NS_YELLOW = "NS_YELLOW"
    EW_GREEN  = "EW_GREEN"
    EW_YELLOW = "EW_YELLOW"


# ─── ANSI 색상 코드 ───────────────────────────────────────────
class Color:
    RED    = "\033[91m"
    YELLOW = "\033[93m"
    GREEN  = "\033[92m"
    BLUE   = "\033[94m"
    CYAN   = "\033[96m"
    BOLD   = "\033[1m"
    DIM    = "\033[2m"
    RESET  = "\033[0m"


# ─── 신호 렌더 헬퍼 ──────────────────────────────────────────
SIGNAL_SYMBOLS = {
    Signal.RED:    (Color.RED,    "●"),
    Signal.YELLOW: (Color.YELLOW, "●"),
    Signal.GREEN:  (Color.GREEN,  "●"),
}

DIM_SYMBOLS = {
    Signal.RED:    (Color.DIM, "○"),
    Signal.YELLOW: (Color.DIM, "○"),
    Signal.GREEN:  (Color.DIM, "○"),
}


def render_light(sig: Signal, active: bool) -> str:
    color, symbol = (SIGNAL_SYMBOLS if active else DIM_SYMBOLS)[sig]
    return f"{color}{symbol}{Color.RESET}"


# ─── 단계 설정 테이블 ─────────────────────────────────────────
@dataclass
class PhaseConfig:
    north: Signal
    south: Signal
    east:  Signal
    west:  Signal
    ped_ns: str   # 'walk' | 'stop'
    ped_ew: str
    label: str


PHASE_CONFIGS: Dict[Phase, PhaseConfig] = {
    Phase.NS_GREEN: PhaseConfig(
        north=Signal.GREEN, south=Signal.GREEN,
        east=Signal.RED,    west=Signal.RED,
        ped_ns="stop",  ped_ew="walk",
        label="남북 직진 (초록)",
    ),
    Phase.NS_YELLOW: PhaseConfig(
        north=Signal.YELLOW, south=Signal.YELLOW,
        east=Signal.RED,     west=Signal.RED,
        ped_ns="stop",  ped_ew="stop",
        label="남북 주의 (황색)",
    ),
    Phase.EW_GREEN: PhaseConfig(
        north=Signal.RED,   south=Signal.RED,
        east=Signal.GREEN,  west=Signal.GREEN,
        ped_ns="walk",  ped_ew="stop",
        label="동서 직진 (초록)",
    ),
    Phase.EW_YELLOW: PhaseConfig(
        north=Signal.RED,    south=Signal.RED,
        east=Signal.YELLOW,  west=Signal.YELLOW,
        ped_ns="stop",  ped_ew="stop",
        label="동서 주의 (황색)",
    ),
}

PHASE_ORDER = [
    Phase.NS_GREEN,
    Phase.NS_YELLOW,
    Phase.EW_GREEN,
    Phase.EW_YELLOW,
]


# ─── 신호등 시스템 ────────────────────────────────────────────
@dataclass
class TrafficLightSystem:
    ns_green_dur:  int = 30
    ns_yellow_dur: int = 5
    ew_green_dur:  int = 30
    ew_yellow_dur: int = 5
    speed:         float = 1.0
    max_cycles:    int = 0   # 0 = 무한 반복

    phase_index: int = field(default=0, init=False)
    cycle_count: int = field(default=0, init=False)
    log: list = field(default_factory=list, init=False)

    def get_duration(self, phase: Phase) -> int:
        return {
            Phase.NS_GREEN:  self.ns_green_dur,
            Phase.NS_YELLOW: self.ns_yellow_dur,
            Phase.EW_GREEN:  self.ew_green_dur,
            Phase.EW_YELLOW: self.ew_yellow_dur,
        }[phase]

    def current_phase(self) -> Phase:
        return PHASE_ORDER[self.phase_index]

    def next_phase(self):
        self.phase_index = (self.phase_index + 1) % len(PHASE_ORDER)
        if self.phase_index == 0:
            self.cycle_count += 1

    def render_frame(self, phase: Phase, time_left: int) -> str:
        cfg = PHASE_CONFIGS[phase]
        lines = []

        # ── 헤더 ──
        lines.append("")
        lines.append(f"{Color.BOLD}{Color.CYAN}  사거리 신호등 시스템  {Color.RESET}")
        lines.append(f"  사이클: {Color.BOLD}{self.cycle_count:3d}{Color.RESET}  "
                     f"단계: {Color.BOLD}{cfg.label}{Color.RESET}  "
                     f"남은시간: {Color.YELLOW}{Color.BOLD}{time_left:3d}초{Color.RESET}")
        lines.append("")

        # ── 교차로 ASCII 다이어그램 ──
        # 신호 표시 함수
        def sig(direction: str) -> str:
            s = getattr(cfg, direction)
            return render_light(s, True)

        def ped(side: str) -> str:
            status = getattr(cfg, f"ped_{side}")
            if status == "walk":
                return f"{Color.GREEN}▶{Color.RESET}"
            else:
                return f"{Color.RED}✖{Color.RESET}"

        # 신호등 수직형 렌더 (R/Y/G)
        def vlight(direction: str) -> list[str]:
            """수직 신호등 3줄 반환"""
            s = getattr(cfg, direction)
            r = render_light(Signal.RED,    s == Signal.RED)
            y = render_light(Signal.YELLOW, s == Signal.YELLOW)
            g = render_light(Signal.GREEN,  s == Signal.GREEN)
            return [r, y, g]

        # 수평형 신호등 (동서용)
        def hlight(direction: str) -> str:
            s = getattr(cfg, direction)
            r = render_light(Signal.RED,    s == Signal.RED)
            y = render_light(Signal.YELLOW, s == Signal.YELLOW)
            g = render_light(Signal.GREEN,  s == Signal.GREEN)
            return f"{r} {y} {g}"

        n_r, n_y, n_g = vlight("north")
        s_r, s_y, s_g = vlight("south")

        w_ped = ped("ew")
        e_ped = ped("ew")
        n_ped = ped("ns")
        s_ped = ped("ns")

        # 교차로 그림
        road = f"{Color.DIM}══{Color.RESET}"
        cross = f"{Color.DIM}╬{Color.RESET}"

        lines += [
            f"               북 (N) {n_ped} ",
            f"               [{n_r}]         ",
            f"               [{n_y}]         ",
            f"               [{n_g}]         ",
            f"               │           ",
            f"  {hlight('west')}  ══{cross}══  {hlight('east')} ",
            f"  서(W){w_ped}          {e_ped}동(E)  ",
            f"               │           ",
            f"               [{s_g}]         ",
            f"               [{s_y}]         ",
            f"               [{s_r}]         ",
            f"               남 (S) {s_ped} ",
        ]

        lines.append("")

        # ── 단계 타임라인 ──
        timeline = []
        for i, p in enumerate(PHASE_ORDER):
            dur = self.get_duration(p)
            label_map = {
                Phase.NS_GREEN:  "NS초록",
                Phase.NS_YELLOW: "NS황색",
                Phase.EW_GREEN:  "EW초록",
                Phase.EW_YELLOW: "EW황색",
            }
            lbl = label_map[p]
            if p == phase:
                timeline.append(f"{Color.CYAN}[{Color.BOLD}{lbl}{dur:2d}s{Color.RESET}{Color.CYAN}]{Color.RESET}")
            else:
                timeline.append(f"{Color.DIM}[{lbl}{dur:2d}s]{Color.RESET}")
        lines.append("  " + " → ".join(timeline))
        lines.append("")

        # ── 보행자 범례 ──
        lines.append(f"  보행자: {Color.GREEN}▶ 통행가능{Color.RESET}  {Color.RED}✖ 정지{Color.RESET}")
        lines.append("")

        return "\n".join(lines)

    def run(self):
        phase = self.current_phase()
        time_left = self.get_duration(phase)

        print("\033[2J\033[H", end="")  # clear screen
        print(f"{Color.DIM}종료: Ctrl+C{Color.RESET}")

        try:
            while True:
                # 화면 갱신
                print("\033[H", end="")  # cursor to top
                frame = self.render_frame(phase, time_left)
                print(frame, end="", flush=True)

                time.sleep(1.0 / self.speed)
                time_left -= 1

                if time_left <= 0:
                    self.next_phase()
                    phase = self.current_phase()
                    time_left = self.get_duration(phase)

                    if self.max_cycles > 0 and self.cycle_count >= self.max_cycles:
                        print(f"\n{Color.BOLD}{Color.GREEN}  {self.max_cycles}사이클 완료. 종료합니다.{Color.RESET}\n")
                        break

        except KeyboardInterrupt:
            print(f"\n\n{Color.DIM}  시뮬레이션 종료.{Color.RESET}\n")
            self._print_summary()

    def _print_summary(self):
        total = (
            self.ns_green_dur + self.ns_yellow_dur +
            self.ew_green_dur + self.ew_yellow_dur
        )
        print(f"{Color.BOLD}=== 시뮬레이션 요약 ==={Color.RESET}")
        print(f"  완료 사이클: {self.cycle_count}")
        print(f"  1사이클 시간: {total}초")
        print(f"  NS 초록: {self.ns_green_dur}s  NS 황색: {self.ns_yellow_dur}s")
        print(f"  EW 초록: {self.ew_green_dur}s  EW 황색: {self.ew_yellow_dur}s")
        print()


# ─── CLI ──────────────────────────────────────────────────────
def main():
    parser = argparse.ArgumentParser(
        description="사거리 신호등 시뮬레이터",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
예시:
  python traffic_light.py                  # 기본 실행
  python traffic_light.py --speed 5        # 5배속
  python traffic_light.py --ns 45 --ew 30  # 남북 45초, 동서 30초
  python traffic_light.py --cycles 2       # 2사이클 실행 후 종료
        """
    )
    parser.add_argument("--ns",     type=int, default=30, metavar="초",
                        help="남북 초록 신호 시간 (기본: 30)")
    parser.add_argument("--ew",     type=int, default=30, metavar="초",
                        help="동서 초록 신호 시간 (기본: 30)")
    parser.add_argument("--ny",     type=int, default=5,  metavar="초",
                        help="남북 황색 신호 시간 (기본: 5)")
    parser.add_argument("--ey",     type=int, default=5,  metavar="초",
                        help="동서 황색 신호 시간 (기본: 5)")
    parser.add_argument("--speed",  type=float, default=1.0,
                        help="시뮬레이션 배속 (기본: 1.0)")
    parser.add_argument("--cycles", type=int, default=0,
                        help="실행 사이클 수 (기본: 0=무한)")

    args = parser.parse_args()

    system = TrafficLightSystem(
        ns_green_dur=args.ns,
        ns_yellow_dur=args.ny,
        ew_green_dur=args.ew,
        ew_yellow_dur=args.ey,
        speed=args.speed,
        max_cycles=args.cycles,
    )

    system.run()


if __name__ == "__main__":
    main()
